data/data-pipeline/data_pipeline/comparison_tool/src/utils.py

```diff
@@ -0,0 +1,356 @@
+    """Reads standardized csvs in
+
+    Parameters:
+        file_path: the file to read
+        columns: the columns to include
+        geoid: the geoid column name (if we change this in usa.csv, will need to change this slightly)
+
+    Returns:
+        dataframe that has been read in with geographic index
+    """
+    assert (
+        geoid == field_names.GEOID_TRACT_FIELD
+    ), f"Field name specified for geoid is incorrect. Use {field_names.GEOID_TRACT_FIELD}"
+    return pd.read_csv(
+        file_path, usecols=columns + [geoid], dtype={geoid: str}
+    ).set_index(geoid)
+
+
+def produce_summary_stats(
+    joined_df: pd.DataFrame,
+    comparator_column: str,
+    score_column: str,
+    population_column: str,
+    geoid_column: str = field_names.GEOID_TRACT_FIELD,
+) -> pd.DataFrame:
+    """Produces high-level overview dataframe
+
+    Parameters:
+        joined_df: the big df
+        comparator_column: the column name for the comparator identification bool
+        score_column: the column name for the CEJST score bool
+        population_column: the column that includes population count per tract
+        geoid_column: the geoid10_tract column
+
+    Returns:
+        population: the high-level overview df
+    """
+    # I actually think this makes sense to be out of all tracts
+    temp_joined_df = joined_df.fillna({comparator_column: "missing"})
+    
+    population_df = temp_joined_df.groupby([comparator_column, score_column]).agg(
+        {population_column: ["sum"], geoid_column: ["count"]}
+    )
+
+    population_df["share_of_tracts"] = (
+        population_df[geoid_column] / population_df[geoid_column].sum()
+    )
+
+    population_df["share_of_population_in_tracts"] = (
+        population_df[population_column]
+        / population_df[population_column].sum()
+    )
+
+    population_df.columns = [
+        "Population",
+        "Count of tracts",
+        "Share of tracts",
+        "Share of population",
+    ]
+    return population_df
+
+
+def get_demo_series(
+    grouping_column: str,
+    joined_df: pd.DataFrame,
+    demo_columns: list,
+) -> pd.DataFrame:
+    """Helper function to produce demographic information"""
+    # To preserve np.nan, we drop all nans
+    full_df = joined_df.dropna(subset=[grouping_column])
+    return (
+        full_df[full_df[grouping_column]][demo_columns]
+        .mean()
+        .T.rename(grouping_column)
+    )
+
+
+def get_tract_level_grouping(
+    joined_df: pd.DataFrame,
+    score_column: str,
+    comparator_column: str,
+    demo_columns: list,
+    keep_missing_values: bool = True,
+) -> pd.DataFrame:
+    """Function to produce segmented statistics (tract level)
+
+    Here, we are thinking about the following segments:
+    1. CEJST and comparator
+    2. Not CEJST and comparator
+    3. Not CEJST and not comparator
+    4. CEJST and not comparator
+
+    If "keep_missing_values" flag:
+    5. Missing from CEJST and comparator (this should never be true!)
```

Comment:
Why would that never be true?

Reply:
because we force all tracts to have a value here -- so it would be a census tract that we've never seen. Missing, not "not a DAC".